[WASM] CheckBox.Unchecked fired ~1 minute *after* it was removed from the visual tree/DOM

### Current behavior

This is weird one. I don't have a repo at this time and I doubt I will be able to create one, since the case is so weird that I suspect it would involved multiple 'obscure' conditions to trigger.

I'm tring to describe the case:
- I have a few `CheckBox `instances on a `StackPanel` which ultimately is hosted by a **Uno.UI.Toolkit.ElevatedView**. Basically I'm simulating a mode dialog (I don't think that matters here, but anyway...)
- as I remove the `StackPanel `from the visual tree I will see that `.Unchecked` for the first `CheckBox` child on the `StackPanel` is fired (I confirmed by tracing)
- the `CheckBox `is not bound to anything but just listed the the `Children` collection. 
- it appears I only could reproduce the problem if the `CheckBox` was checked as the 'dialog' was closed. If feels like there would be some logic in place: "OK, this thing is closed. Now there is a checkbox which is/was checked. We have to uncheck it to clear things up." (pure speculation obviously)

Here is a snippet:
```
new CheckBox
{
HorizontalAlignment     = HorizontalAlignment.Center,
IsChecked                      = Common.Session.Current.IsExpertMode,
Margin                           = new Thickness(24, 8, 0, 0),
VerticalAlignment          = VerticalAlignment.Center,
Visibility                         = (Gui.MainPage.IsPortraitMode ? Visibility.Collapsed : Visibility.Visible),
}
```
.. there really is nothing extraordinary.

What's going on there? How could an event for `CheckBox `ever be fired after it's removed from the visual tree/DOM???

### Expected behavior

_No response_

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

No

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

@Youssef1313 I get this error when trying to compile
```
4>C:\Users\dierk\Documents\unusual-src\Unusual.Gui.Shared\Gui\MainPage.cs(2429,1,2429,38): error CS0234: The type or namespace name 'FrameworkTemplatePool' does not exist in the namespace 'Windows.UI.Xaml' (are you missing an assembly reference?)

```

--- a/src/SamplesApp/SamplesApp.Shared/App.xaml.cs
+++ b/src/SamplesApp/SamplesApp.Shared/App.xaml.cs
@@ -522,11 +522,6 @@ namespace SamplesApp
 #if __SKIA__
 			Uno.UI.FeatureConfiguration.ToolTip.UseToolTips = true;
 #endif
-
-#if HAS_UNO
-			// Allow template pool to work under higher memory load for CI.
-			FrameworkTemplatePool.HighMemoryThreshold = 0.9f;
-#endif
 		}
 
 

--- a/src/Uno.UI.RuntimeTests/Helpers/FeatureConfigurationHelper.cs
+++ b/src/Uno.UI.RuntimeTests/Helpers/FeatureConfigurationHelper.cs
@@ -9,6 +9,13 @@ namespace Uno.UI.RuntimeTests.Helpers
 {
 	public static class FeatureConfigurationHelper
 	{
+#if !NETFX_CORE
+		private class MockProvider : FrameworkTemplatePoolDefaultPlatformProvider
+		{
+			public override bool CanUseMemoryManager => false;
+		}
+#endif
+
 		/// <summary>
 		/// Enable <see cref="FrameworkTemplate"/> pooling (Uno only) for the duration of a single test.
 		/// </summary>
@@ -19,7 +26,12 @@ namespace Uno.UI.RuntimeTests.Helpers
 #else
 			var originallyEnabled = FrameworkTemplatePool.IsPoolingEnabled;
 			FrameworkTemplatePool.IsPoolingEnabled = true;
-			return Disposable.Create(() => FrameworkTemplatePool.IsPoolingEnabled = originallyEnabled);
+			FrameworkTemplatePool.Instance.SetPlatformProvider(new MockProvider());
+			return Disposable.Create(() =>
+			{
+				FrameworkTemplatePool.IsPoolingEnabled = originallyEnabled;
+				FrameworkTemplatePool.Instance.SetPlatformProvider(null);
+			});
 #endif
 		}
 

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs
@@ -133,11 +133,6 @@ namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml_Controls
 			}
 		}
 
-		/// <remarks>
-		/// This test is memory sensitive - if the memory usage exceeds FrameworkTemplatePool.HighMemoryThreshold,
-		/// the test will most likely faily as the template will be recycled. Currently the Samples app uses
-		/// HighMemoryThreshold of 0.9 (see App.ConfigureFeatureFlags).
-		/// </remarks>
 		[TestMethod]
 #if __MACOS__
 		[Ignore("Currently fails on macOS, part of #9282 epic")]

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_FrameworkTemplatePool.cs
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_FrameworkTemplatePool.cs
@@ -1,0 +1,101 @@
+﻿using System;
+using System.Threading.Tasks;
+using Uno.UI.RuntimeTests.Helpers;
+using Windows.UI.Xaml.Controls;
+using static Private.Infrastructure.TestServices;
+
+namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml_Controls;
+
+[TestClass]
+[RunsOnUIThread]
+internal class Given_FrameworkTemplatePool
+{
+	[TestMethod]
+	public async Task TestCheckBox()
+	{
+		using (FeatureConfigurationHelper.UseTemplatePooling())
+		{
+			var scrollViewer = new ScrollViewer();
+			var elevatedViewChild = new Uno.UI.Toolkit.ElevatedView();
+			scrollViewer.Content = elevatedViewChild;
+
+			var c = new CheckBox();
+			c.IsChecked = true;
+			bool uncheckedFired = false;
+			c.Unchecked += (_, _) => uncheckedFired = true;
+			elevatedViewChild.ElevatedContent = c;
+
+			WindowHelper.WindowContent = scrollViewer;
+			await WindowHelper.WaitForLoaded(scrollViewer);
+			var template = scrollViewer.Template;
+
+			scrollViewer.Template = null;
+			scrollViewer.Template = template;
+
+			scrollViewer.ApplyTemplate();
+
+			Assert.IsFalse(uncheckedFired);
+		}
+	}
+
+	[TestMethod]
+	public async Task TestTextBox()
+	{
+		using (FeatureConfigurationHelper.UseTemplatePooling())
+		{
+			var scrollViewer = new ScrollViewer();
+
+			var textBox = new TextBox();
+			bool textChangedFired = false;
+			textBox.TextChanged += (_, _) => textChangedFired = true;
+			textBox.Text = "Text";
+
+			// TextBox dispatches raising the event. We wait here until the above TextChanged is raised.
+			await Windows.ApplicationModel.Core.CoreApplication.MainView.Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () => { });
+
+			Assert.IsTrue(textChangedFired);
+			textChangedFired = false;
+
+			scrollViewer.Content = textBox;
+
+			WindowHelper.WindowContent = scrollViewer;
+			await WindowHelper.WaitForLoaded(scrollViewer);
+			var template = scrollViewer.Template;
+
+			scrollViewer.Template = null;
+			scrollViewer.Template = template;
+
+			scrollViewer.ApplyTemplate();
+
+			await Windows.ApplicationModel.Core.CoreApplication.MainView.Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () => { });
+
+			Assert.IsFalse(textChangedFired);
+		}
+	}
+
+	[TestMethod]
+	public async Task TestToggleSwitch()
+	{
+		using (FeatureConfigurationHelper.UseTemplatePooling())
+		{
+			var scrollViewer = new ScrollViewer();
+
+			var toggleSwitch = new ToggleSwitch();
+			toggleSwitch.IsOn = true;
+			bool toggledFired = false;
+			toggleSwitch.Toggled += (_, _) => toggledFired = true;
+			scrollViewer.Content = toggleSwitch;
+
+			WindowHelper.WindowContent = scrollViewer;
+			await WindowHelper.WaitForLoaded(scrollViewer);
+			var template = scrollViewer.Template;
+
+			scrollViewer.Template = null;
+			scrollViewer.Template = template;
+
+			scrollViewer.ApplyTemplate();
+
+			Assert.IsFalse(toggledFired);
+		}
+	}
+}

--- a/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs
@@ -9,6 +9,12 @@ namespace Windows.UI.Xaml.Controls.Primitives
 	public partial class ToggleButton : ButtonBase, IFrameworkTemplatePoolAware
 	{
 		/// <summary>
+		/// This is a workaround for template pooling issue where we change IsChecked when template is recycled.
+		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// </summary>
+		private bool _suppressCheckedChanged;
+
+		/// <summary>
 		/// Initializes a new instance of the ToggleButton class.
 		/// </summary>
 		public ToggleButton()
@@ -82,6 +88,11 @@ namespace Windows.UI.Xaml.Controls.Primitives
 
 		protected virtual void OnIsCheckedChanged(bool? oldValue, bool? newValue)
 		{
+			if (_suppressCheckedChanged)
+			{
+				return;
+			}
+
 			if (IsChecked == null)
 			{
 				// Indeterminate
@@ -99,7 +110,18 @@ namespace Windows.UI.Xaml.Controls.Primitives
 			}
 		}
 
-		public void OnTemplateRecycled() => IsChecked = false;
+		public void OnTemplateRecycled()
+		{
+			try
+			{
+				_suppressCheckedChanged = true;
+				IsChecked = false;
+			}
+			finally
+			{
+				_suppressCheckedChanged = false;
+			}
+		}
 
 		internal void AutomationPeerToggle() => OnClick();
 	}

--- a/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs
@@ -11,6 +11,7 @@ namespace Windows.UI.Xaml.Controls.Primitives
 		/// <summary>
 		/// This is a workaround for template pooling issue where we change IsChecked when template is recycled.
 		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 		/// </summary>
 		private bool _suppressCheckedChanged;
 
@@ -88,6 +89,7 @@ namespace Windows.UI.Xaml.Controls.Primitives
 
 		protected virtual void OnIsCheckedChanged(bool? oldValue, bool? newValue)
 		{
+			// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 			if (_suppressCheckedChanged)
 			{
 				return;

--- a/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs
@@ -9,8 +9,8 @@ namespace Windows.UI.Xaml.Controls.Primitives
 	public partial class ToggleButton : ButtonBase, IFrameworkTemplatePoolAware
 	{
 		/// <summary>
-		/// This is a workaround for template pooling issue where we change IsChecked when template is recycled.
-		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// This is a workaround for the template pooling issue where we change IsChecked when the template is recycled.
+		/// This prevents incorrect event raising but is not a "real" solution. Pooling could still cause issues.
 		/// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 		/// </summary>
 		private bool _suppressCheckedChanged;

--- a/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.mux.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.mux.cs
@@ -227,6 +227,12 @@ namespace Windows.UI.Xaml.Controls.Primitives
 		/// </summary>
 		private void OnIsCheckedChanged()
 		{
+			if (_suppressCheckedChanged)
+			{
+				UpdateVisualState();
+				return;
+			}
+
 			var isChecked = IsChecked;
 			if (isChecked == null)
 			{

--- a/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.mux.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.mux.cs
@@ -227,7 +227,7 @@ namespace Windows.UI.Xaml.Controls.Primitives
 		/// </summary>
 		private void OnIsCheckedChanged()
 		{
-			if (_suppressCheckedChanged)
+			if (_suppressCheckedChanged) // Uno workaround
 			{
 				UpdateVisualState();
 				return;

--- a/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.mux.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.mux.cs
@@ -227,6 +227,7 @@ namespace Windows.UI.Xaml.Controls.Primitives
 		/// </summary>
 		private void OnIsCheckedChanged()
 		{
+			// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 			if (_suppressCheckedChanged) // Uno workaround
 			{
 				UpdateVisualState();

--- a/src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs
@@ -43,6 +43,12 @@ namespace Windows.UI.Xaml.Controls
 
 	public partial class TextBox : Control, IFrameworkTemplatePoolAware
 	{
+		/// <summary>
+		/// This is a workaround for template pooling issue where we change IsChecked when template is recycled.
+		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// </summary>
+		private bool _suppressTextChanged;
+
 #pragma warning disable CS0067, CS0649
 		private IFrameworkElement _placeHolder;
 		private ContentControl _contentElement;
@@ -304,13 +310,17 @@ namespace Windows.UI.Xaml.Controls
 			{
 				_isInvokingTextChanged = true;
 				_isTextChangedPending = false;
-				TextChanged?.Invoke(this, new TextChangedEventArgs(this));
+				if (!_suppressTextChanged)
+				{
+					TextChanged?.Invoke(this, new TextChangedEventArgs(this));
+				}
 			}
 #if !HAS_EXPENSIVE_TRYFINALLY // Try/finally incurs a very large performance hit in mono-wasm - https://github.com/dotnet/runtime/issues/50783
 			finally
 #endif
 			{
 				_isInvokingTextChanged = false;
+				_suppressTextChanged = false;
 			}
 		}
 
@@ -1050,6 +1060,7 @@ namespace Windows.UI.Xaml.Controls
 
 		public void OnTemplateRecycled()
 		{
+			_suppressTextChanged = true;
 			Text = string.Empty;
 		}
 

--- a/src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs
@@ -44,8 +44,8 @@ namespace Windows.UI.Xaml.Controls
 	public partial class TextBox : Control, IFrameworkTemplatePoolAware
 	{
 		/// <summary>
-		/// This is a workaround for template pooling issue where we change IsChecked when template is recycled.
-		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// This is a workaround for the template pooling issue where we change IsChecked when the template is recycled.
+		/// This prevents incorrect event raising but is not a "real" solution. Pooling could still cause issues.
 		/// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 		/// </summary>
 		private bool _suppressTextChanged;

--- a/src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs
@@ -46,6 +46,7 @@ namespace Windows.UI.Xaml.Controls
 		/// <summary>
 		/// This is a workaround for template pooling issue where we change IsChecked when template is recycled.
 		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 		/// </summary>
 		private bool _suppressTextChanged;
 
@@ -310,7 +311,7 @@ namespace Windows.UI.Xaml.Controls
 			{
 				_isInvokingTextChanged = true;
 				_isTextChangedPending = false;
-				if (!_suppressTextChanged)
+				if (!_suppressTextChanged) // This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 				{
 					TextChanged?.Invoke(this, new TextChangedEventArgs(this));
 				}

--- a/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.cs
@@ -14,6 +14,12 @@ namespace Windows.UI.Xaml.Controls
 	public partial class ToggleSwitch : Control, IFrameworkTemplatePoolAware
 	{
 		/// <summary>
+		/// This is a workaround for template pooling issue where we change IsOn when template is recycled.
+		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// </summary>
+		private bool _suppressToggled;
+
+		/// <summary>
 		/// Initializes a new instance of the ToggleSwitch class.
 		/// </summary>
 		public ToggleSwitch()
@@ -221,7 +227,18 @@ namespace Windows.UI.Xaml.Controls
 			}
 		}
 
-		public void OnTemplateRecycled() => IsOn = false;
+		public void OnTemplateRecycled()
+		{
+			try
+			{
+				_suppressToggled = true;
+				IsOn = false;
+			}
+			finally
+			{
+				_suppressToggled = false;
+			}
+		}
 
 		internal void AutomationPeerToggle() => IsOn = !IsOn;
 	}

--- a/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.cs
@@ -16,6 +16,7 @@ namespace Windows.UI.Xaml.Controls
 		/// <summary>
 		/// This is a workaround for template pooling issue where we change IsOn when template is recycled.
 		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 		/// </summary>
 		private bool _suppressToggled;
 

--- a/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.cs
@@ -14,8 +14,8 @@ namespace Windows.UI.Xaml.Controls
 	public partial class ToggleSwitch : Control, IFrameworkTemplatePoolAware
 	{
 		/// <summary>
-		/// This is a workaround for template pooling issue where we change IsOn when template is recycled.
-		/// This prevents incorrect event raising, but is not a "real" solution. Pooling could still cause issues.
+		/// This is a workaround for the template pooling issue where we change IsOn when the template is recycled.
+		/// This prevents incorrect event raising but is not a "real" solution. Pooling could still cause issues.
 		/// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 		/// </summary>
 		private bool _suppressToggled;

--- a/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.mux.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.mux.cs
@@ -419,6 +419,7 @@ namespace Windows.UI.Xaml.Controls
 			var args = new RoutedEventArgs();
 			args.OriginalSource = this;
 
+			// This workaround can be removed if pooling is removed. See https://github.com/unoplatform/uno/issues/12189
 			if (!_suppressToggled) // Uno workaround.
 			{
 				Toggled?.Invoke(this, args);

--- a/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.mux.cs
+++ b/src/Uno.UI/UI/Xaml/Controls/ToggleSwitch/ToggleSwitch.mux.cs
@@ -419,7 +419,10 @@ namespace Windows.UI.Xaml.Controls
 			var args = new RoutedEventArgs();
 			args.OriginalSource = this;
 
-			Toggled?.Invoke(this, args);
+			if (!_suppressToggled) // Uno workaround.
+			{
+				Toggled?.Invoke(this, args);
+			}
 
 			if (!_isDragging)
 			{

--- a/src/Uno.UI/UI/Xaml/FrameworkTemplatePoolDefaultPlatformProvider.cs
+++ b/src/Uno.UI/UI/Xaml/FrameworkTemplatePoolDefaultPlatformProvider.cs
@@ -31,7 +31,7 @@ class FrameworkTemplatePoolDefaultPlatformProvider : IFrameworkTemplatePoolPlatf
 	public TimeSpan Now
 		=> _watch.Elapsed;
 
-	public bool CanUseMemoryManager
+	public virtual bool CanUseMemoryManager
 		=> _canUseMemoryManager;
 
 	public ulong AppMemoryUsage

